source/opt/inline_pass.cpp

```diff
@@ -270,30 +319,120 @@ std::unique_ptr<BasicBlock> InlinePass::AddGuardBlock(
   return new_blk_ptr;
 }
 
+DebugScope InlinePass::BuildDebugScope(const DebugScope& callee_instr_scope,
+                                       DebugInlinedAtContext& inlined_at_ctx) {
+  return DebugScope(callee_instr_scope.GetLexicalScope(),
+                    BuildDebugInlinedAtChain(callee_instr_scope.GetInlinedAt(),
+                                             inlined_at_ctx));
+}
+
+uint32_t InlinePass::BuildDebugInlinedAtChain(
+    uint32_t callee_inlined_at, DebugInlinedAtContext& inlined_at_ctx) {
+  if (inlined_at_ctx.FunctionCallScope().GetLexicalScope() == kNoDebugScope)
+    return kNoInlinedAt;
+
+  // Reuse the already generated DebugInlinedAt chain if exists.
+  uint32_t already_generated_chain_head_id =
+      inlined_at_ctx.InlinedAtChainHead(callee_inlined_at);
+  if (already_generated_chain_head_id != kNoInlinedAt) {
+    return already_generated_chain_head_id;
+  }
+
+  const uint32_t inlined_to_caller =
+      context()->get_debug_info_mgr()->CreateDebugInlinedAt(
+          inlined_at_ctx.FunctionCallLine(),
+          inlined_at_ctx.FunctionCallScope());
+  if (inlined_to_caller == kNoInlinedAt) return kNoInlinedAt;
+
+  // Simply specify that this instruction was inlined to the caller
+  // if it has no DebugInlinedAt in advance.
+  if (callee_inlined_at == kNoInlinedAt) {
+    // Keep the new chain information that will be reused it.
```

Comment:
This comment would not be needed if the `UpdateInlinedAtContext` was named better.